library/core/src/any.rs

```diff
@@ -594,10 +663,36 @@ impl dyn Any + Send + Sync {
 /// While `TypeId` implements `Hash`, `PartialOrd`, and `Ord`, it is worth
 /// noting that the hashes and ordering will vary between Rust releases. Beware
 /// of relying on them inside of your code!
-#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+#[cfg_attr(not(bootstrap), lang = "TypeId")]
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
 #[stable(feature = "rust1", since = "1.0.0")]
 pub struct TypeId {
-    t: u64,
+    hash: u64,
+
+    // NOTE(eddyb) this follows `hash`, so that the auto-derived `PartialEq`
+    // doesn't waste time comparing `mangling`, when `hash` differs.
```

Comment:
Please document all the considerations here. Basically we want a quick reject/accept in all the cases, so

* `eq`
    * can return `false` if the hashes differ
    * can return `true` if the pointers are the same
    * only if neither of these did anything do we need to check the actual data behind the pointer
* `ne`
    * can return `true` if the hashes differ
    * can return `false` if the pointers are the same
    * only if neither of these did anything do we need to check the actual data behind the pointer